// coc-markmap/dist/index.common.js
/*! coc-markmap v0.4.4 | MIT License */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var path = require('path');
var coc_nvim = require('coc.nvim');
var markmapCli = require('markmap-cli');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e['default'] : e; }

var path__default = /*#__PURE__*/_interopDefaultLegacy(path);

let devServer;

async function getFullText() {
  const doc = await coc_nvim.workspace.document;
  return doc.textDocument.getText();
}

async function getSelectedText() {
  const doc = await coc_nvim.workspace.document;
  const range = await coc_nvim.workspace.getSelectedRange('v', doc);
  return doc.textDocument.getText(range);
}

async function getRangeText(line1, line2) {
  const {
    nvim
  } = coc_nvim.workspace;
  const lines = await nvim.eval(`getline(${line1},${line2})`);
  return lines.join('\n');
}

async function createMarkmapFromVim(content, options) {
  const {
    nvim
  } = coc_nvim.workspace;
  const input = await nvim.eval('expand("%:p")');

  if (options.watch) {
    if (devServer) devServer.close();
    devServer = await markmapCli.develop({
      input,
      open: true
    });
  } else {
    const basename = path__default.basename(input, path__default.extname(input));
    markmapCli.createMarkmap({ ...options,
      content,
      output: basename && `${basename}.html`,
      open: true
    });
  }
}

function activate(context) {
  // const config = workspace.getConfiguration('markmap');
  context.subscriptions.push(coc_nvim.workspace.registerKeymap(['n'], 'markmap-create', async () => {
    await createMarkmapFromVim(await getFullText());
  }, {
    sync: false
  }));
  context.subscriptions.push(coc_nvim.workspace.registerKeymap(['v'], 'markmap-create-v', async () => {
    await createMarkmapFromVim(await getSelectedText());
  }, {
    sync: false
  }));
  context.subscriptions.push(coc_nvim.commands.registerCommand('markmap.create', async (...args) => {
    const positional = [];
    const options = {};

    for (const arg of args) {
      if (['-w', '--watch'].includes(arg)) options.watch = true;else if (!arg.startsWith('-')) positional.push(arg);
    }

    const [line1, line2] = positional;
    const content = line1 && line2 ? await getRangeText(line1, line2) : await getFullText();
    await createMarkmapFromVim(content, options);
  }));
}

exports.activate = activate;
